File: node/todo.js
(function () {
    const fs = require('fs');
    const path = require('path');
    const readline = require('readline');

    const argv = process.argv;

    const dbPath = __dirname + path.sep + 'db';

    const verb = argv[2];
    const n = parseInt(argv[3]) - 1;
    const content = argv[3];
    const n1 = parseInt(argv[4]);
    const content1 = argv[4];

    const tips = {
        no_content_err: '请填写任务内容。',
        task_no_exist_err: '任务不存在。',
        num_err: '清输入合法的任务序号。',
        step_err: '清输入合法的移动步数。',
        origin_content_is: '原内容是：',
        please_input_content: '请输入要修改的内容：',
        your_verb_is: '你的动作是： ',
        i_dont_know_what_your_want: '我不知道你想干什么~',
        are_you_sure_clear_all: '确认清空所有任务吗？[Y/n]',
    };

    var list = [];

    var noAction = false;

    init();

    switch (verb) {
        case 'add':
            add(content);
            break;

        case 'list':
            break;

        case 'done':
            done(n);
            break;

        case 'delete':
            del(n);
            break;

        case 'edit':
            (async () => {
                await edit(n, content1);
                save();
                display();
            })();
            break;

        case 'undone':
            undone(n);
            break;

        case 'moveup':
            moveup(n, n1);
            break;

        case 'movedown':
            movedown(n, n1);
            break;

        case 'clearall':
            (async () => {
                await clearall();
                save();
            })();
            break;

        default:
            noAction = true;
            console.log(tips.your_verb_is + verb);
            console.log(tips.i_dont_know_what_your_want);
            break;
    }

    if (!['list', 'edit', 'clearall'].includes(verb) && !noAction) save();

    if (!['edit', 'clearall'].includes(verb)) display();


    /* ================ helper ================ */


    /**
     * 初始化。
     */
    function init() {
        if (fs.existsSync(dbPath)) {
            list = fs.readFileSync(dbPath, {encoding: 'utf8'});

            try {
                list = JSON.parse(list);
            } catch (e) {
                // ...
            }
        } else {
            fs.writeFileSync(dbPath, '[]');
        }
    }

    /**
     * 保存数据。
     */
    function save() {
        fs.writeFileSync(dbPath, JSON.stringify(list));
    }

    /**
     * 显示任务列表。
     */
    function display() {
        console.log('');
        // 保证列表中每一项任务的完成标记对齐。
        let idxLen = String(list.length).length;
        list.forEach((item, idx) => {
            idx = String(idx + 1);
            let space = ' '.repeat(idxLen - idx.length);
            let status = item.status ? '[x]' : '[_]';
            let content = item.content;

            console.log(space + idx + '. ' + status + ' ' + content);
        });
    }

    /**
     * 添加任务。
     * @param content 任务内容。
     */
    function add(content) {
        if (checkContent(content)) list.push({content, status: false});
    }

    /**
     * 标记任务为完成。
     * @param n 要标记的任务序号。
     */
    function done(n) {
        if (checkNum(n) && checkTaskExist(n, list)) list[n].status = true;
    }

    /**
     * 删除任务。
     * @param n 要删除的任务序号。
     */
    function del(n) {
        if (checkNum(n) && checkTaskExist(n, list)) list.splice(n, 1);
    }

    /**
     * 编辑任务内容，如果命令行参数里面没有带任务内容，则后续过程中询问内容。
     * @param n 要编辑的任务序号。
     * @param content 命令行参数中获取到的任务内容。
     * @return {Promise<void>}
     */
    async function edit(n, content) {
        if (checkNum(n) && checkTaskExist(n, list)) {
            const task = list[n];

            if (typeof content === 'undefined') {
                console.log(tips.origin_content_is, task.content);
                task.content = await ask(tips.please_input_content);
            } else {
                task.content = content;
            }
        }
    }

    /**
     * 标记任务为已完成。
     * @param n 要标记为完成的任务序号。
     */
    function undone(n) {
        if (checkNum(n) && checkTaskExist(n, list)) list[n].status = false;
    }

    /**
     * 向上移动任务。
     * @param n 要向上移动的任务序号。
     * @param n1 要移动的步数。
     */
    function moveup(n, n1) {
        if (checkNum(n) && checkStep(n1) && checkTaskExist(n, list)) {
            let idx = n - n1;
            idx = idx < 0 ? 0 : idx;
            let task = list.splice(n, 1);
            list.splice(idx, 0, ...task);
        }
    }

    /**
     * 向下移动任务。
     * @param n 要向下移动的任务序号。
     * @param n1 要移动的步数。
     */
    function movedown(n, n1) {
        if (checkNum(n) && checkStep(n1) && checkTaskExist(n, list)) {
            let idx = n + n1;
            idx = idx > list.length ? list.length : idx;
            let task = list.splice(n, 1);
            list.splice(idx, 0, ...task);
        }
    }

    /**
     * 清除所有任务。
     * @return {Promise<boolean>}
     */
    async function clearall() {
        let answer = await ask(tips.are_you_sure_clear_all);

        answer = answer.toLowerCase();

        if (['n', 'no'].includes(answer)) return false;
        else if (['y', 'yes'].includes(answer)) {
            list = [];
            return true;
        }
    }


    /* ================ common func ================ */


    /**
     * 检查任务内容是否填写。
     * @param content 任务内容。
     * @return {boolean} true
     */
    function checkContent(content) {
        if (typeof content === 'undefined') {
            console.log(tips.no_content_err);
            process.exit(0);
        }

        return true;
    }

    /**
     * 检查任务序号是否合法。
     * @param n 任务序号。
     * @return {boolean} true
     */
    function checkNum(n) {
        if (Number.isNaN(n)) {
            console.log(tips.num_err);
            process.exit(0);
        }

        return true;
    }

    /**
     * 检查列表中是否存在对应序号的任务。
     * @param n 任务序号。
     * @param list 任务列表。
     * @return {boolean} true
     */
    function checkTaskExist(n, list) {
        if (typeof list[n] === 'undefined') {
            console.log(tips.task_no_exist_err);
            process.exit(0);
        }

        return true;
    }

    /**
     * 检查移动步数是否合法。
     * @param n1 移动步数。
     * @return {boolean} true
     */
    function checkStep(n1) {
        if (Number.isNaN(n1)) {
            console.log(tips.step_err);
            process.exit(0);
        }

        return true;
    }

    /**
     * 命令行提问互动。
     * @param question 问题。
     * @return {Promise<Promise>}  异步 Promise 对象。
     */
    function ask(question) {
        return new Promise(resolve => {
            const rl = readline.createInterface({
                input: process.stdin,
                output: process.stdout,
                prompt: question
            });

            rl.prompt();

            rl.on('line', (line) => {
                if (0 === line.length) rl.prompt();
                else {
                    resolve(line);
                    rl.close();
                }
            });
        });
    }
})();

File: python/todo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import json
import os
import re
import sys

TIPS = {
    'no_content_err': '请填写任务内容。',
    'task_no_exist_err': '任务不存在。',
    'num_err': '清输入合法的任务序号。',
    'step_err': '清输入合法的移动步数。',
    'origin_content_is': '原内容是：',
    'please_input_content': '请输入要修改的内容：',
    'your_verb_is': '你的动作是： ',
    'i_dont_know_what_your_want': '我不知道你想干什么~',
    'are_you_sure_clear_all': '确认清空所有任务吗？[N/y]',
}

DB_PATH = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'db')

# Ｔｏｄｏ 类。


class Todo(object):

    # 构造对象、初始化并调用指令。
    def __init__(self):

        # 类似 JavaScript 的 parseInt 函数。
        def parseInt(s):
            try:
                return int(re.match(r'^\d+', s).group(0))
            except:
                return 0

        args = sys.argv

        try:
            self.__verb = args[1]
        except:
            self.__verb = None
        try:
            self.__n = parseInt(args[2]) - 1
        except:
            self.__n = None
        try:
            self.__content = args[2]
        except:
            self.__content = None
        try:
            self.__n1 = parseInt(args[3])
        except:
            self.__n1 = None
        try:
            self.__content1 = args[3]
        except:
            self.__content1 = None

        # 任务列表。
        self.__list = []
        # 是否为无效动作。
        self.__noAction = False
        # 命令行的输入参数是否验证通过，当输入有误的情况下，不打印列表，不保存列表。
        self.__validated = True

        self.__init()

        self.__exec()

    # 销毁对象前打印任务列表并保存。
    def __del__(self):
        if self.__validated:
            if self.__verb not in ['list', 'clearall']:
                self.__save()
            if self.__verb not in ['clearall']:
                self.__display()

    # 初始化，获取数据库数据。
    def __init(self):
        try:
            with open(DB_PATH, 'r', encoding='utf-8') as f:
                self.__list = json.load(f)
        except FileNotFoundError:
            with open(DB_PATH, 'w', encoding='utf-8') as f:
                json.dump([], f)
        except json.decoder.JSONDecodeError:
            self.__list = []

    # 保存数据到数据库。
    def __save(self):
        with open(DB_PATH, 'w', encoding='utf-8') as f:
            json.dump(self.__list, f)

    # 执行操作。
    def __exec(self):
        verb = self.__verb

        content = self.__content
        content1 = self.__content1
        n = self.__n
        n1 = self.__n1
        if verb == 'list':
            pass
        elif verb == 'add':     # 添加任务。
            if self.__checkContent(content):
                self.__list.append({"content": content, "status": False})
        elif verb == 'done':    # 标记任务为完成。
            if self.__checkNum(n) and self.__checkTaskExist(n, self.__list):
                self.__list[n]['status'] = True
        elif verb == 'delete':  # 删除任务。
            if self.__checkNum(n) and self.__checkTaskExist(n, self.__list):
                self.__list.pop(n)
        elif verb == 'edit':    # 编辑任务内容，如果命令行参数里面没有带任务内容，则后续过程中询问内容。
            if self.__checkNum(n) and self.__checkTaskExist(n, self.__list):
                if content1 != None and len(content1) > 0:
                    pass
                else:
                    print(TIPS['origin_content_is'] +
                          self.__list[n]['content'])
                    while content1 == None or len(content1) <= 0:
                        content1 = input(TIPS['please_input_content'])
                self.__list[n]['content'] = content1
        elif verb == 'undone':  # 标记任务为已完成。
            if self.__checkNum(n) and self.__checkTaskExist(n, self.__list):
                self.__list[n]['status'] = False
        elif verb == 'moveup':  # 向上移动任务。
            if self.__checkNum(n) and self.__checkStep(n1) and self.__checkTaskExist(n, self.__list):
                idx = n - n1
                idx = 0 if idx < 0 else idx
                tasks = self.__list.pop(n)
                self.__list.insert(idx, tasks)
        elif verb == 'movedown':    # 向下移动任务。
            if self.__checkNum(n) and self.__checkStep(n1) and self.__checkTaskExist(n, self.__list):
                idx = n + n1
                listLen = len(self.__list)
                idx = listLen if idx > listLen else idx
                tasks = self.__list.pop(n)
                self.__list.insert(idx, tasks)
        elif verb == 'clearall':    # 清除所有任务。
            answer = None
            while answer == None or len(answer) <= 0:
                answer = input(TIPS['are_you_sure_clear_all']).strip().lower()
            if answer in ['n', 'no']:
                pass
            elif answer in ['y', 'yes']:
                self.__list = []
                self.__save()
        else:
            self.__noAction = True
            print(TIPS['your_verb_is'] + verb)
            print(TIPS['i_dont_know_what_your_want'])

    # 显示任务列表。
    def __display(self):
        if self.__noAction:
            return

        print('')

        idxLen = len(str(len(self.__list)))
        for idx, item in enumerate(self.__list):
            idx = str(idx + 1)
            space = ' ' * (idxLen - len(idx))
            status = '[x]' if item['status'] else '[_]'
            content = item['content']
            print(space + idx + '.', status, content)

    # 检查任务内容是否填写。
    def __checkContent(self, content):
        if not content:
            self.__validated = False
            print(TIPS['no_content_err'])
            sys.exit()
        return True

    # 检查任务序号是否合法。
    def __checkNum(self, n):
        if n == None or not n >= 0:
            self.__validated = False
            print(TIPS['num_err'])
            sys.exit()
        return True

    # 检查列表中是否存在对应序号的任务。
    def __checkTaskExist(self, n, list):
        if n >= len(list):
            self.__validated = False
            print(TIPS['task_no_exist_err'])
            sys.exit()
        return True

    # 检查移动步数是否合法。
    def __checkStep(self, n1):
        if n1 == None or not n1 >= 0:
            self.__validated = False
            print(TIPS['step_err'])
            sys.exit()
        return True


if __name__ == '__main__':
    Todo()
